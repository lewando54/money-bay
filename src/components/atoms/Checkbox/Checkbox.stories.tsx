/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { useState } from 'react'
import React from 'react'
import { Meta,StoryObj } from '@storybook/react-native'

import Checkbox from './Checkbox'

import { GOOD_GREEN } from '@styling/GlobalTheme.style'

interface IMockupCheckboxProps {
	checked?: boolean
    checkedbackgroundColor?: string
    circleBackgroundColor?: string
    backgroundColor?: string
}

function StatefulCheckbox({
    checked,
    checkedbackgroundColor,
    circleBackgroundColor,
    backgroundColor
}: IMockupCheckboxProps) {
    const [checkedState, setCheckedState] = useState(checked)

    return (
        <Checkbox 
            checked={checkedState} 
            testID='test'
            backgroundColor={backgroundColor}
            circleBackgroundColor={circleBackgroundColor}
            checkedbackgroundColor={checkedbackgroundColor}
            onPress={(newValue) => {setCheckedState(newValue)}}
        />
    )
}

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
    component: StatefulCheckbox,
    // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
    argTypes: {
        checked: { control: 'boolean' },
        checkedbackgroundColor: { control: 'color' },
        circleBackgroundColor: { control: 'color' },
        backgroundColor: { control: 'color' },
    }
} satisfies Meta<typeof StatefulCheckbox>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
export default meta
type Story = StoryObj<typeof StatefulCheckbox>

export const Default: Story = {
    args: {
        checked: false,
        checkedbackgroundColor: GOOD_GREEN,
        circleBackgroundColor: 'white',
        backgroundColor: '#CED6E1',
    }
}
// More on args: https://storybook.js.org/docs/react/writing-stories/args
