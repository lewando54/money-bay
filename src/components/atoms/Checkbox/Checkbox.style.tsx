import { StyleSheet } from 'react-native'
export default StyleSheet.create({
    container: {
        width: 39.48,
        height: 24,
        justifyContent: 'center',
        borderRadius: 20,
        padding: 2
    },
    circle: {
        width: 20.9,
        height: 20.9,
        backgroundColor: 'white',
        borderRadius: 20.9
    },
    shadow: {
        elevation: 5,
        shadowOffset: {width: 0, height: 5},
        shadowRadius: 8,
        shadowColor: '#000',
        shadowOpacity: 0.15
    }
})
