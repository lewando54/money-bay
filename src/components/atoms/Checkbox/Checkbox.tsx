import React from 'react'
import { Pressable } from 'react-native'
import Animated, { Easing,interpolateColor, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'

import checkboxStyle from './Checkbox.style'

import { GOOD_GREEN } from '@styling/GlobalTheme.style'

interface ICheckboxProps {
    checked: boolean
    checkedbackgroundColor?: string
    circleBackgroundColor?: string
    backgroundColor?: string
    onPress: (newValue: boolean) => void
    testID?: string
}

export default function Checkbox({
    checked=false,
    checkedbackgroundColor=GOOD_GREEN,
    circleBackgroundColor='white',
    backgroundColor='#CED6E1',
    onPress,
    testID
}: ICheckboxProps){

    const translateX = useSharedValue(0)
    const backgroundColorProgress = useSharedValue(0)

    const animatedCircleStyle = useAnimatedStyle(() => ({
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        transform: [{ translateX: withTiming(translateX.value * 2, {duration: 500, easing: Easing.inOut(Easing.exp)}) }],
    }))

    const animatedBackgroundColorStyle = useAnimatedStyle(() => ({
        backgroundColor: interpolateColor(
            backgroundColorProgress.value,
            [0, 1],
            [backgroundColor, checkedbackgroundColor]
        )
    }))

    const handleOnPress = () => {
        onPress(!checked)
        
        if(!checked){
            translateX.value += 7
        }
        else{
            translateX.value -= 7
        }

        backgroundColorProgress.value = withTiming(1 - backgroundColorProgress.value, { duration: 500 })
    }

    return (
        <Animated.View style={[checkboxStyle.container, animatedBackgroundColorStyle]}>
            <Pressable onPress={handleOnPress} testID={testID}>
                <Animated.View style={[checkboxStyle.circle, {backgroundColor: circleBackgroundColor}, checkboxStyle.shadow, animatedCircleStyle]}></Animated.View>
            </Pressable>
        </Animated.View>
    )
}
