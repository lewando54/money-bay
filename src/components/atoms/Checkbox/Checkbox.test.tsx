import React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render, screen } from '@testing-library/react-native'
import Checkbox from './Checkbox'

describe('Checkbox', () => {
  jest.useFakeTimers();
  it('should render correctly', () => {
    const tree = renderer.create(<Checkbox checked={false} onPress={jest.fn()}/>)
    expect(tree).toMatchSnapshot()
  })

  it('should be checked when pressed', () => {
    const setChecked = jest.fn()
    const { getByTestId } = render(<Checkbox checked={false} onPress={(newValue) => setChecked(newValue)} testID='test'/>)
    fireEvent.press(getByTestId('test'))
    expect(setChecked).toHaveBeenCalledWith(true)
  })

  it('should be unchecked when pressed while checked', () => {
    const setChecked = jest.fn()
    const { getByTestId } = render(<Checkbox checked={true} onPress={(newValue) => setChecked(newValue)} testID='test'/>)
    fireEvent.press(getByTestId('test'))
    expect(setChecked).toHaveBeenCalledWith(false)
  })

  it('should call onPress when pressed', () => {
    const handleOnPress = jest.fn()
    const { getByTestId } = render(<Checkbox checked={false} onPress={handleOnPress} testID='test'/>)
    fireEvent.press(getByTestId('test'))
    expect(handleOnPress).toHaveBeenCalled()
  })
})
