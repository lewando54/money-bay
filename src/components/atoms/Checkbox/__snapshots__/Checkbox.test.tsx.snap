// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "backgroundColor": "rgba(206, 214, 225, 1)",
      },
    }
  }
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(206, 214, 225, 1)",
      "borderRadius": 20,
      "height": 24,
      "justifyContent": "center",
      "padding": 2,
      "width": 39.48,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      animatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        {
          "backgroundColor": "white",
          "borderRadius": 20.9,
          "elevation": 5,
          "height": 20.9,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 5,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 8,
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 20.9,
        }
      }
    />
  </View>
</View>
`;
